tcms/issuetracker/types.py

```diff
@@ -384,6 +386,99 @@ def report_issue_from_testcase(self, caserun):
         return url + '/issues/new?' + urlencode(args, True)
 
 
+class Redmine(IssueTrackerType):
+    """
+        Support for Redmine. Requires:
+
+        :api_url: - the API URL for your Redmine instance
+        :api_username: - a username registered in Redmine
+        :api_password: - the password for this username
+    """
+
+    def __init__(self, tracker):
+        super().__init__(tracker)
+
+        if not self.is_adding_testcase_to_issue_disabled():
+            self.rpc = redminelib.Redmine(
+                tracker.api_url,
+                username=self.tracker.api_username,
+                password=self.tracker.api_password
+            )
+
+    def add_testcase_to_issue(self, testcases, issue):
+        for case in testcases:
+            redmine_integration.RedmineThread(self.rpc, case, issue).start()
+
+    def all_issues_link(self, ids):
+        if not self.tracker.base_url:
+            return None
+
+        if not self.tracker.base_url.endswith('/'):
+            self.tracker.base_url += '/'
+
+        query = 'issues?utf8=✓&set_filter=1&sort=id%3Adesc&f%5B%5D=issue_id'
+        query += '&op%5Bissue_id%5D=%3D&v%5Bissue_id%5D%5B%5D={0}'.format('%2C'.join(ids))
+        query += '&f%5B%5D=&c%5B%5D=tracker&c%5B%5D=status&c%5B%5D=priority'
+        query += '&c%5B%5D=subject&c%5B%5D=assigned_to&c%5B%5D=updated_on&group_by=&t%5B%5D='
+
+        return self.tracker.base_url + query
+
+    def find_project_by_name(self, caserun):
+        projects = self.rpc.project.all()
+        for prj in projects:
+            if prj.name == caserun.run.plan.product.name:
+                project = prj
+                break
+
+        if not project:
+            return None
+
+        return project
+
+    @staticmethod
+    def find_tracker_by_name_with_default(project):
+        tracker = project.trackers[0]
+        for trk in project.trackers:
+            if trk == 'Bug':
+                tracker = trk
+                break
+
+        return tracker
+
+    def report_issue_from_testcase(self, caserun):
+        # note: Redmine instance needs to have the same projects
+        # defined otherwise this will fail!
+        project = self.find_project_by_name(caserun)
+
+        if not project:
+            return self.tracker.base_url
+
+        tracker = self.find_tracker_by_name_with_default(project)
+
+        query = "issue[{0}]=".format('tracker_id') + quote_plus(str(tracker.id))
+        query += "&issue[{0}]=".format('subject')
+        query += "{0}".format(quote_plus('Failed test:{0}'.format(caserun.case.summary)))
```

Comment:
Why use `format` here? Looks like you can directly append the return value of `quote_plus()`.

Reply:
Why do we need `quote_plus` instead of just `quote` ?

Reply:
Line 458 will be changed to 
query = "issue[tracker_id]=" + str(tracker.id)

And all of the **quote_plus** will be replaced with **quote**